### ISSUE TITLE
Adversarial training for feature invariance?

### ISSUE THREAD

**Issue by GilesStrong** [GilesStrong/lumin#75]
So far I've been contacted by two researchers about LUMIN, both of whom were wanting to use it for adversarial training in order to get invariance to some feature of the data. Unfortunately, LUMIN does not offer this out of the box, but seeing as 100% of the people that were interested in using LUMIN are wanting to do adversarial training, it's perhaps something that we should aim to offer in the future.

I made some attempt at this, but never tested it and don't have suitable data to hand to really develop it myself. I have a feeling, though, that a decent implementation will probably be quite task-specific, so input from fellow researchers more experience with this is really necessary to move forward.

My initial attempt is pasted below, along with a short response to an email:

"however it is potentially implementable by modifying the training loop and inheriting from the Model class.

In the attached files there's an example of this, in which the new training loop (adversarial_fold_train_ensemble) takes two ModelBuilders, one to provide the primary models and the other to provide the adversarial models. Two sets of callbacks get created, one for each model. A new Model class (AdversarialModel) inherits from Model and has a new method (adversarial_fit) which takes the adversarial model as an argument. This is can then be modified to provide interaction between the two models as necessary. Does this look like the kind of thing which could help? I'm afraid I haven't had a chance to test it, but I've marked everything new with "# New".

In its current form, adversarial_fold_train_ensemble will only save the primary model, and this is determined by the validation loss returned by primary_model.evaluate, so you may want to modify this to avoid saving models which don't have a flat response."

```
from typing import Dict, List, Tuple, Any, Optional
from pathlib import Path
from fastprogress import master_bar, progress_bar
import pickle
import timeit
import numpy as np
import os
import sys
from random import shuffle
from collections import OrderedDict
import math
from functools import partial
import warnings

import torch.tensor as Tensor

from lumin.nn.data.fold_yielder import FoldYielder
from lumin.nn.data.batch_yielder import BatchYielder
from lumin.nn.models.model_builder import ModelBuilder
from lumin.nn.models.model import Model
from lumin.nn.callbacks.cyclic_callbacks import AbsCyclicCallback
from lumin.nn.callbacks.model_callbacks import AbsModelCallback
from lumin.nn.callbacks.abs_callback import AbsCallback
from lumin.utils.misc import to_tensor, to_device
from lumin.utils.statistics import uncert_round
from lumin.nn.metrics.eval_metric import EvalMetric
from lumin.plotting.training import plot_train_history
from lumin.plotting.plot_settings import PlotSettings
from lumin.nn.training.metric_logger import MetricLogger
from lumin.nn.models.abs_model import AbsModel

import matplotlib.pyplot as plt

__all__ = ['adversarial_fold_train_ensemble', 'AdversarialModel']


# New___
class AdversarialModel(Model):
    def adversarial_fit(self, batch_yielder:BatchYielder, adversary_model:AbsModel, primary_callbacks:Optional[List[AbsCallback]]=None,
                        adversary_callbacks:Optional[List[AbsCallback]]=None, mask_inputs:bool=True) -> float:
        r'''
        Fit network for one complete iteration of a :class:`~lumin.nn.data.batch_yielder.BatchYielder`, i.e. one (sub-)epoch

        Arguments:
            batch_yielder: :class:`~lumin.nn.data.batch_yielder.BatchYielder` providing training data in form of tuple of inputs, targets,
                and weights as tensors on device
            adversary_model: :class:`~lumin.nn.models.model.Model` to act as the adversarial model during training
            primary_callbacks: list of :class:`~lumin.nn.callbacks.abs_callback.AbsCallback` to be used during training of the primary model
            adversary_callbacks: list of :class:`~lumin.nn.callbacks.abs_callback.AbsCallback` to be used during training of the adversary model
            mask_inputs: whether to apply input mask if one has been set

        Returns:
            Loss on training data averaged across all minibatches
        '''

        self.model.train()
        adversary_model.model.train()
        self.stop_train = False
        if primary_callbacks   is None: primary_callbacks   = []
        if adversary_callbacks is None: adversary_callbacks = []

        for c in primary_callbacks:   c.on_epoch_begin(by=batch_yielder)
        for c in adversary_callbacks: c.on_epoch_begin(by=batch_yielder)

        if self.input_mask is not None and mask_inputs: batch_yielder.inputs = batch_yielder.inputs[:,self.input_mask]

        # Replace this as necessary
        # _________________________
        # losses = []
        # for x, y, w in batch_yielder:
        #     for c in callbacks: c.on_batch_begin()
        #     y_pred = self.model(x)
        #     loss = self.loss(weight=w)(y_pred, y) if w is not None else self.loss()(y_pred, y)
        #     losses.append(loss.data.item())
        #     self.opt.zero_grad()
        #     for c in callbacks: c.on_backwards_begin(loss=loss)
        #     loss.backward()
        #     for c in callbacks: c.on_backwards_end(loss=loss)
        #     self.opt.step()
            
        #     for c in callbacks: c.on_batch_end(loss=losses[-1])
        #     if self.stop_train: break
        
        # for c in callbacks: c.on_epoch_end(losses=losses)
        # return np.mean(losses)
        # _________________________
# ______


def _get_folds(val_idx, n_folds, shuffle_folds:bool=True):
    r'''
    Return (shuffled) list of fold indeces which does not include the validation index
    '''

    folds = [x for x in range(n_folds) if x != val_idx]
    if shuffle_folds: shuffle(folds)
    return folds


def adversarial_fold_train_ensemble(fy:FoldYielder, n_models:int, bs:int, primary_model_builder:ModelBuilder, adversarial_model_builder:ModelBuilder,
                                    callback_partials:Optional[List[partial]]=None, eval_metrics:Optional[Dict[str,EvalMetric]]=None,
                                    train_on_weights:bool=True, eval_on_weights:bool=True, patience:int=10, max_epochs:int=200,
                                    shuffle_fold:bool=True, shuffle_folds:bool=True, bulk_move:bool=True,
                                    live_fdbk:bool=True, live_fdbk_first_only:bool=True, live_fdbk_extra:bool=True, live_fdbk_extra_first_only:bool=False,
                                    savepath:Path=Path('train_weights'), verbose:bool=False, log_output:bool=False,
                                    plot_settings:PlotSettings=PlotSettings(), plots:Optional[Any]=None) \
        -> Tuple[List[Dict[str,float]],List[Dict[str,List[float]]],List[Dict[str,float]]]:
    r'''
    Adversarial training method for :class:`~lumin.nn.models.model.Model`.
    Trains a specified numer of models created by a :class:`~lumin.nn.models.model_builder.ModelBuilder` on data provided by a
    :class:`~lumin.nn.data.fold_yielder.FoldYielder`, and save them to savepath.
    Note, this does not return trained models, instead they are saved and must be loaded later. Instead this method returns results of model training.
    Each :class:`~lumin.nn.models.model.Model` is trained on N-1 folds, for a :class:`~lumin.nn.data.fold_yielder.FoldYielder` with N folds, and the remaining
    fold is used as validation data.
    Training folds are loaded iteratively, and model evaluation takes place after each fold use (a sub-epoch), rather than after ever use of all folds (epoch).
    Training continues until:
        - All of the training folds are used max_epoch number of times;
        - Or validation loss does not decrease for patience number of training folds;
          (or cycles, if using an :class:`~lumin.nn.callbacks.cyclic_callbacks.AbsCyclicCallback`);
        - Or a callback triggers trainign to stop, e.g. :class:`~lumin.nn.callbacks.cyclic_callbacks.OneCycle`
        
    Depending on the live_fdbk arguments, live plots of losses and other metrics may be shown during training, if running in Jupyter. By default, a live plot
    with extra information will be shown for training the first model, and afterwards no live plots will be shown. Shoing the live plot slightly slows down the
    training, but can help highlight problems without having to wait to the end. Thererfore this compromises between showing useful information and training
    speed, since any problems should hopefully be visible in the first model.

    Once training is finished, the state with the lowest validation loss is loaded, evaluated, and saved.

    Arguments:
        fy: :class:`~lumin.nn.data.fold_yielder.FoldYielder` interfacing ot training data
        n_models: number of models to train
        bs: batch size. Number of data points per iteration
        primary_model_builder: :class:`~lumin.nn.models.model_builder.ModelBuilder` creating the primary networks to train
        adversarial_model_builder: :class:`~lumin.nn.models.model_builder.ModelBuilder` creating the adversary networks
        callback_partials: optional list of functools.partial, each of which will a instantiate :class:`~lumin.nn.callbacks.callback.Callback` when called
        eval_metrics: list of instantiated :class:`~lumin.nn.metric.eval_metric.EvalMetric`.
            At the end of training, validation data and model predictions will be passed to each, and the results printed and saved
        train_on_weights: If weights are present in training data, whether to pass them to the loss function during training
        eval_on_weights: If weights are present in validation data, whether to pass them to the loss function during validation
        patience: number of folds (sub-epochs) or cycles to train without decrease in validation loss before ending training (early stopping)
        max_epochs: maximum number of epochs for which to train
        live_fdbk: whether or not to show any live feedback at all during training (slightly slows down training, but helps spot problems)
        live_fdbk_first_only: whether to only show live feedback for the first model trained (trade off between time and problem spotting)
        live_fdbk_extra: whether to show extra information live feedback (further slows training)
        live_fdbk_extra_first_only: whether to only show extra live feedback information for the first model trained (trade off between time and information)
        shuffle_fold: whether to tell :class:`~lumin.nn.data.batch_yielder.BatchYielder` to shuffle data
        shuffle_folds: whether to shuffle the order of the trainign folds
        bulk_move: whether to pass all training data to device at once, or by minibatch. Bulk moving will be quicker, but may not fit in memory.
        savepath: path to to which to save model weights and results
        verbose: whether to print out extra information during training
        log_output: whether to save printed results to a log file rather than printing them
        plot_settings: :class:`~lumin.plotting.plot_settings.PlotSettings` class to control figure appearance
        plots: Depreciated: loss history will always be shown,
            lr history will no longer be shown separately,
            and live feedback is now controlled by `live_fdbk` argument

    Returns:
        - results list of validation losses and other eval_metrics results, ordered by model training.
            Can be used to create an :class:`~lumin.nn.ensemble.ensemble.Ensemble`.
        - histories list of loss histories, ordered by model training
        - cycle_losses if an :class:`~lumin.nn.callbacks.cyclic_callbacks.AbsCyclicCallback` was passed, list of validation losses at the end of each cycle,
            ordered by model training. Can be passed to :class:`~lumin.nn.ensemble.ensemble.Ensemble`.
    '''

    os.makedirs(savepath, exist_ok=True)
    os.system(f"rm {savepath}/*.h5 {savepath}/*.json {savepath}/*.pkl {savepath}/*.png {savepath}/*.log")
    if callback_partials is None: callback_partials = []
    
    if log_output:
        old_stdout = sys.stdout
        log_file = open(savepath/'training_log.log', 'w')
        sys.stdout = log_file

    if plots is not None:
        warnings.warn("The plots argument is now depreciated and ignored. Loss history will always be shown, lr history will no longer be shown separately, \
                       and live feedback is now controlled by the four live_fdbk arguments. This argument will be removed in V0.6.")

    train_tmr = timeit.default_timer()
    results,histories,cycle_losses = [],[],[]
    nb = len(fy.foldfile['fold_0/targets'])//bs

    if live_fdbk:
        metric_log = MetricLogger(loss_names=['Train', 'Validation'], n_folds=fy.n_folds, extra_detail=live_fdbk_extra or live_fdbk_extra_first_only,
                                  plot_settings=plot_settings)
    
    model_bar = master_bar(range(n_models))
    for model_num in (model_bar):
        model_bar.show()
        val_id = model_num % fy.n_folds
        print(f"Training model {model_num+1} / {n_models}, Val ID = {val_id}")
        if model_num == 1:
            if live_fdbk_first_only: live_fdbk = False  # Only show fdbk for first training
            elif live_fdbk_extra_first_only: metric_log.extra_detail = False
        if live_fdbk: metric_log.reset()
        model_tmr = timeit.default_timer()
        os.system(f"rm {savepath}/best.h5")
        best_loss,epoch_counter,subEpoch,stop = math.inf,0,0,False
        loss_history = OrderedDict({'trn_loss': [], 'val_loss': []})
        cycle_losses.append({})
        trn_ids = _get_folds(val_id, fy.n_folds, shuffle_folds)

        primary_model,adversary_model = AdversarialModel(primary_model_builder),AdversarialModel(adversarial_model_builder)  # New

        val_fold = fy.get_fold(val_id)
        if not eval_on_weights: val_fold['weights'] = None

        primary_cyclic_callback,primary_callbacks,primary_loss_callbacks = None,[],[]
        adversary_cyclic_callback,adversary_callbacks = None,[]  # New

        for c in callback_partials: primary_callbacks.append(c(model=primary_model))
        for c in primary_callbacks:
            if isinstance(c, AbsCyclicCallback):
                c.set_nb(nb)
                primary_cyclic_callback = c
        for c in primary_callbacks:
            if isinstance(c, AbsModelCallback):
                c.set_val_fold(val_fold)
                c.set_cyclic_callback(primary_cyclic_callback)
                if getattr(c, "get_loss", None):
                    primary_loss_callbacks.append(c)
                    if live_fdbk: metric_log.add_loss_name(type(c).__name__)
                    loss_history[f'{type(c).__name__}_val_loss'] = []
        for c in primary_callbacks: c.on_train_begin(model_num=model_num, savepath=savepath)

        # New___
        for c in callback_partials: adversary_callbacks.append(c(model=adversary_model))
        for c in adversary_callbacks:
            if isinstance(c, AbsCyclicCallback):
                c.set_nb(nb)
                adversary_cyclic_callback = c
        for c in adversary_callbacks:
            if isinstance(c, AbsModelCallback):
                c.set_val_fold(val_fold)
                c.set_cyclic_callback(adversary_cyclic_callback)
        for c in adversary_callbacks: c.on_train_begin(model_num=model_num, savepath=savepath)
        # ______

        # Validation data
        if bulk_move:
            if fy.has_matrix and fy.yield_matrix: val_x = (to_device(Tensor(val_fold['inputs'][0]).float()), to_device(Tensor(val_fold['inputs'][1]).float())) 
            else:                                 val_x =  to_device(Tensor(val_fold['inputs']).float())
            val_y = to_device(Tensor(val_fold['targets'])) if bulk_move else Tensor(val_fold['targets'])
            if train_on_weights: val_w = to_device(to_tensor(val_fold['weights'])) if bulk_move else to_tensor(val_fold['weights'])
            else:                val_w = None
            if 'multiclass' in primary_model_builder.objective: val_y = val_y.long().squeeze()
            else:                                       val_y = val_y.float()

        epoch_pb = progress_bar(range(max_epochs), leave=True)
        if live_fdbk: model_bar.show()
        for epoch in epoch_pb:
            for trn_id in trn_ids:
                subEpoch += 1
                batch_yielder = BatchYielder(**fy.get_fold(trn_id), objective=primary_model_builder.objective,
                                             bs=bs, use_weights=train_on_weights, shuffle=shuffle_fold, bulk_move=bulk_move)
                loss_history['trn_loss'].append(primary_model.adversarial_fit(batch_yielder, primary_callbacks=primary_callbacks,
                                                adversary_model=adversary_model, adversary_callbacks=adversary_callbacks))  # New
                del batch_yielder

                if bulk_move:
                    val_loss = primary_model.evaluate(val_x, val_y, weights=val_w, callbacks=primary_callbacks)
                else:
                    batch_yielder = BatchYielder(**val_fold, objective=primary_model_builder.objective,
                                                 bs=bs, use_weights=train_on_weights, shuffle=shuffle_fold, bulk_move=bulk_move)
                    val_loss = primary_model.evaluate_from_by(batch_yielder, callbacks=primary_callbacks)
                    del batch_yielder

                loss_history['val_loss'].append(val_loss)
                loss_callback_idx = None
                loss = val_loss
                for i, lc in enumerate(primary_loss_callbacks):
                    l = lc.get_loss()
                    if l < loss: loss, loss_callback_idx = l, i
                    if verbose: print(f'{subEpoch} {type(lc).__name__} loss {l}, default loss {val_loss}')
                    l = loss if l is None or not lc.active else l
                    loss_history[f'{type(lc).__name__}_val_loss'].append(l)

                if primary_cyclic_callback is not None and primary_cyclic_callback.cycle_end:
                    if verbose: print(f"Saving snapshot {primary_cyclic_callback.cycle_count}")
                    cycle_losses[-1][primary_cyclic_callback.cycle_count] = val_loss
                    primary_model.save(str(savepath/f"{model_num}_cycle_{primary_cyclic_callback.cycle_count}.h5"))

                if loss <= best_loss:
                    best_loss = loss
                    epoch_pb.comment = f'Epoch {subEpoch}, best loss: {best_loss:.4E}'
                    if verbose: print(epoch_pb.comment)
                    epoch_counter = 0
                    if loss_callback_idx is not None: primary_loss_callbacks[loss_callback_idx].test_model.save(savepath/"best.h5")
                    else: primary_model.save(savepath/"best.h5")
                elif primary_cyclic_callback is not None:
                    if primary_cyclic_callback.cycle_end: epoch_counter += 1
                else:
                    epoch_counter += 1

                if live_fdbk: metric_log.update_vals([loss_history[l][-1] for l in loss_history])
                if epoch_counter >= patience or primary_model.stop_train:  # Early stopping
                    print('Early stopping after {} epochs'.format(subEpoch))
                    stop = True; break
            if live_fdbk: metric_log.update_plot(best_loss)
            if stop: break

        primary_model.load(savepath/"best.h5")
        primary_model.save(savepath/f'train_{model_num}.h5')
        for c in primary_callbacks: c.on_train_end(fy=fy, val_id=val_id, bs=bs if not bulk_move else None)

        histories.append({})
        histories[-1] = loss_history
        results.append({})
        results[-1]['loss'] = best_loss
        if eval_metrics is not None and len(eval_metrics) > 0:
            y_pred = primary_model.predict(val_fold['inputs'], bs=bs if not bulk_move else None)
            for m in eval_metrics: results[-1][m] = eval_metrics[m].evaluate(fy, val_id, y_pred)
        print(f"Scores are: {results[-1]}")
        with open(savepath/'results_file.pkl', 'wb') as fout: pickle.dump(results, fout)
        with open(savepath/'cycle_file.pkl', 'wb') as fout: pickle.dump(cycle_losses, fout)
        
        plt.clf()
        print(f"Fold took {timeit.default_timer()-model_tmr:.3f}s\n")

    print("\n______________________________________")
    print("Training finished")
    print(f"Cross-validation took {timeit.default_timer()-train_tmr:.3f}s ")
    plot_train_history(histories, savepath/'loss_history', settings=plot_settings)
    for score in results[0]:
        mean = uncert_round(np.mean([x[score] for x in results]), np.std([x[score] for x in results])/np.sqrt(len(results)))
        print(f"Mean {score} = {mean[0]}±{mean[1]}")
    print("______________________________________\n")
    if log_output:
        sys.stdout = old_stdout
        log_file.close()
    return results, histories, cycle_losses
```

**Comment by GilesStrong** [comment]
Closed by recent merge of `PivotTraining` callback
